### ISSUE TITLE
LiteralString cannot be assigned to str

### ISSUE THREAD

**Issue by alexprengere** [astral-sh/ty#4065]
### Summary

This currently passes on mypy and pyright but fails on ty:
```python
def get_str() -> str:
    return "hello world"

a: list[str] = (get_str() or "").split()
```
The error is:
```
error[invalid-assignment]: Object of type `list[str] | list[LiteralString]` is not assignable to `list[str]`
 --> toto.py:5:4
  |
5 | a: list[str] = (get_str() or "").split()
  |    ---------   ^^^^^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `list[str] | list[LiteralString]`
  |    |
  |    Declared type
  |
info: element `list[LiteralString]` of union `list[str] | list[LiteralString]` is not assignable to `list[str]`

Found 1 diagnostic
```
If I follow the advice and replace `list` by `Sequence` then it works.

### Version

0.0.62 (latest)

**Comment by carljm** [comment]
Thanks for the report!

The root cause here is that we infer the type `(str & ~AlwaysFalsy) | Literal[""]` for the expression (`get_str() or ""`), and we don't simplify this back to simply `str`. (Technically I guess there could be a subclass of `str` with custom truthiness behavior, but I think we'd be well advised to ignore that possibility.) Then we distribute this union over the call to `.split()`, and because of the overloads on `split` for literal strings, we get back `list[str]` for the first element of the union (`str & ~AlwaysFalsy`) and `list[LiteralString]` for the second part (`Literal[""]`). Both of those are correct in isolation. The end result is that we infer the awkward union `list[str] | list[LiteralString]` for the overall expression `(get_str() or "").split()`. This union doesn't simplify (due to list invariance) and it also isn't assignable to either `list[str]` or to `list[LiteralString]` (again due to invariance).

There are a few potential fix directions here. Perhaps the simplest one is the proposal in #2233, which would cause the type `(str & ~AlwaysFalsy) | Literal[""]` to collapse to just `str`. Alternatively we could try to recognize the type `(str & ~AlwaysFalsy) | Literal[""]` as simplifiable to `str` in union building, though I expect it would get quite complex to handle all cases here. There may also be ways to handle this via better bidirectional checking (taking advantage of the `list[str]` annotation you have here on the variable.)
